connectors/es/sink.py

```diff
@@ -181,47 +203,86 @@ async def _bulk_api_call():
+                for op, doc in entry.items():
+                    if (
+                        isinstance(doc, dict)
+                        and "_id" in doc.keys()
+                        and "_index" in doc.keys()
+                    ):  # avoiding update bulk extra entries
+                        result[doc["_id"]] = op
+        return result
+
+    async def _process_bulk_response(self, res, ids_to_ops, do_log=False):
         for item in res.get("items", []):
-            if "index" in item:
-                action_item = "index"
-            elif "delete" in item:
-                action_item = "delete"
-            elif "create" in item:
-                action_item = "create"
-            elif "update" in item:
-                action_item = "update"
+            if OP_INDEX in item:
+                action_item = OP_INDEX
+            elif OP_DELETE in item:
+                action_item = OP_DELETE
+            elif OP_CREATE in item:
+                action_item = OP_CREATE
+            elif OP_UPDATE in item:
+                action_item = OP_UPDATE
             else:
                 # Should only happen, if the _bulk API changes
                 # Unlikely, but as this functionality could be used for audits we want to detect changes fast
-                self._logger.error(
-                    f"Unknown action item returned from _bulk API for item {item}"
-                )
+                if do_log:
+                    self._logger.error(
+                        f"Unknown action item returned from _bulk API for item {item}"
+                    )
+                self.counters.increment(OP_UNKNOWN, namespace=BULK_RESPONSES)
                 continue
 
+            self.counters.increment(action_item, namespace=BULK_RESPONSES)
             doc_id = item[action_item].get("_id")
             if doc_id is None:
                 # Should only happen, if the _bulk API changes
                 # Unlikely, but as this functionality could be used for audits we want to detect changes fast
-                self._logger.error(f"Could not retrieve '_id' for document {item}")
+                if do_log:
+                    self._logger.error(f"Could not retrieve '_id' for document {item}")
+                self.counters.increment(ID_MISSING, namespace=BULK_RESPONSES)
                 continue
 
             result = item[action_item].get("result")
-            successful_result = result in SUCCESSFUL_RESULTS
 
+            requested_op = ids_to_ops.get(doc_id, None)
+            if requested_op is None:
+                self.counters.increment(
+                    ID_CHANGED_AFTER_REQUEST, namespace=BULK_RESPONSES
+                )
+            elif action_item != OP_UPDATE and result == "updated":
+                self.counters.increment(ID_DUPLICATE, namespace=BULK_RESPONSES)
+
+            if result == "noop":
+                self.counters.increment(DOCS_DROPPED, namespace=BULK_RESPONSES)
```

Comment:
noop == "dropped by pipeline"?

IMO might be useful to put short code comments here about what's the potential cause of each of that.

Reply:
> noop == "dropped by pipeline"?

I guess so?

Try it out:

```
PUT test

PUT _ingest/pipeline/test-pipeline
{
  "processors": [
    {
      "drop": {
        "if": "true"
      }
    }
  ]
}

POST test/_bulk?pipeline=test-pipeline
{"index": {"_index": "test", "_id":"1"}}
{"foo":"bar"}

```

Results in:

```
{
  "errors": false,
  "took": 0,
  "ingest_took": 0,
  "items": [
    {
      "index": {
        "_index": "test",
        "_id": "1",
        "_version": -3,
        "result": "noop",
        "_shards": {
          "total": 0,
          "successful": 0,
          "failed": 0
        },
        "status": 200
      }
    }
  ]
}
```


> IMO might be useful to put short code comments here about what's the potential cause of each of that.

Can do. 👍

Reply:
side note, thanks for this comment. I went to leave a code comment with a link to the docs, but found that the docs were actually outdated, and didnt' include `noop` or `not_found`. Tech writing team is gonna fix that, now.